### PR TITLE
Adds injectionRate to FaultInjectionServerErrorResult

### PR DESCRIPTION
**Pull request by NaluTripician**
# Description

Adds Application Percentage for ServerErrorRules. This allows users to specify how frequently server error rules are applied.  For instance, by setting this value to `.5` that rule will be applied 50% of the time. This value is 1, or 100%, by default. 

[Mirror](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4450)

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java [hunk 1]
```diff
@@ -14,17 +14,20 @@ public final class FaultInjectionServerErrorResult implements IFaultInjectionRes
...
     private final Boolean suppressServiceRequests;
+    private final Double applyPercentage;
 
     FaultInjectionServerErrorResult(
         FaultInjectionServerErrorType serverErrorTypes,
         Integer times,
         Duration delay,
-        Boolean suppressServiceRequests) {
+        Boolean suppressServiceRequests,
+        Double applyPercentage) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java:24]: Double in Java is Nullable<double> in .Net - please amke this double (non-nullable) with a default of 1 (100 percent)?

Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java:24]: NIT:  find the name applyPercentage really weird. I assume it means failureToBeAppliedPercentage or faultApplicationPercentage?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java:24]: maybe just injectionRate?

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java [hunk 2]
```diff
@@ -65,12 +68,20 @@ public Boolean getSuppressServiceRequests() {
         return this.suppressServiceRequests;
     }
 
+    /***
+     * Get A double between (0,1] representing the percent of times that the rule will be applied.
+     * Default value is 1.0 or 100%
+     * @return the apply percentage.
+     */
+    public Double getApplyPercentage() { return this.applyPercentage; }
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java:76]: getInjectionRate?

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/implementation/faultinjection/FaultInjectionServerErrorResultInternal.java [hunk 3]
```diff
@@ -33,19 +33,22 @@ public class FaultInjectionServerErrorResultInternal {
...
+    private final Double applyPercentage;
 
 
     public FaultInjectionServerErrorResultInternal(
         FaultInjectionServerErrorType serverErrorTypes,
         Integer times,
         Duration delay,
-        Boolean suppressServiceRequests) {
+        Boolean suppressServiceRequests,
+        Double applyPercentage) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/implementation/faultinjection/FaultInjectionServerErrorResultInternal.java:44]: also double

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/implementation/faultinjection/FaultInjectionServerErrorRule.java [hunk 4]
```diff
@@ -32,6 +33,7 @@ public class FaultInjectionServerErrorRule implements IFaultInjectionRuleInterna
     private final FaultInjectionConnectionType connectionType;
     private final FaultInjectionConditionInternal condition;
     private final FaultInjectionServerErrorResultInternal result;
+    private final Random random = new Random();
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/implementation/faultinjection/FaultInjectionServerErrorRule.java:36]: Random in Java 8 and above is thread-safe - can we use a static Random here?

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java [hunk 5]
```diff
@@ -65,12 +68,20 @@ public Boolean getSuppressServiceRequests() {
...
+     * Default value is 1.0 or 100%
+     * @return the apply percentage.
+     */
+    public double getInjectionRate() { return this.injectionRate; }
+
     @Override
     public String toString() {
         return String.format(
-            "FaultInjectionServerErrorResult{ serverErrorType=%s, times=%s, delay=%s }",
+            "FaultInjectionServerErrorResult{ serverErrorType=%s, times=%s, delay=%s, injectionRate=%s }",
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java:81]: should double be logged with %d?

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/implementation/faultinjection/FaultInjectionServerErrorRule.java [hunk 6]
```diff
@@ -108,6 +111,12 @@ public boolean isApplicable(FaultInjectionRequestArgs requestArgs) {
                 String.format("%s [Hit Limit reached. Configured hitLimit %d, evaluationCount %d]", this.id, this.hitLimit, evaluationCount)
             );
             return false;
+        } else if (random.nextDouble() > this.result.getInjectionRate()) {
+            requestArgs.getServiceRequest().faultInjectionRequestContext.recordFaultInjectionRuleEvaluation(
+                requestArgs.getTransportRequestId(),
+                String.format("%s Apply percentage: Rule will not be applied. Configured applyPercentage %f%%", this.id, this.result.getInjectionRate() * 100)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/implementation/faultinjection/FaultInjectionServerErrorRule.java:117]: Change to injection rate.

**Review by kushagraThapar — APPROVED** [review]
Overall looks great @NaluTripician , thank you. 
Please add some more testing cases.

File: sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java (hunk 5 above)
Comment (NaluTripician) [reply sdk/cosmos/azure-cosmos-test/src/main/java/com/azure/cosmos/test/faultinjection/FaultInjectionServerErrorResult.java:81]: Good catch, should be `%f`

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/faultinjection/FaultInjectionServerErrorRuleOnDirectTests.java [hunk 7]
```diff
@@ -1069,6 +1069,50 @@ public void connectionAcquisitionTimeoutAlignConnectionTimeout() throws JsonProc
...
+                cosmosAsyncContainer,
+                Arrays.asList(applyPercentageRule)).block();
+
+            for(int i = 0; i < 100; i++){
+                this.performDocumentOperation(cosmosAsyncContainer, OperationType.Read, createdItem);
+            }
+
+            //Because applyPercentage is based on Random probability,
+            //we expect that this assert will fail 0.66% of the time.
+            assertThat(applyPercentageRule.getHitCount()).isBetween(37L, 63L);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/faultinjection/FaultInjectionServerErrorRuleOnDirectTests.java:1108]: then we should remove the hitLimit config on the rule? else the max hit count should be limited by 3

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks for adding it 👍

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - thanks!

**Pull request merged**